src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -70,589 +91,522 @@ public async ValueTask DisposeAsync()
             }
 
             _disposed = true;
-            _queuedCommandsWriter.Complete();
-            await _pipeWriter.CompleteAsync().ConfigureAwait(false);
+
+            if (_pipeWriter != null)
+                await _pipeWriter.CompleteAsync().ConfigureAwait(false);
+            if (_pipeReader != null)
+                await _pipeReader.CompleteAsync().ConfigureAwait(false);
+
+            await _readerLoopTask.ConfigureAwait(false);
         }
         finally
         {
             _semLock.Release();
         }
     }
 
-    public NatsPipeliningWriteProtocolProcessor CreateNatsPipeliningWriteProtocolProcessor(ISocketConnection socketConnection) => new(socketConnection, this, _opts, _counter);
-
-    public ValueTask ConnectAsync(ClientOpts connectOpts, CancellationToken cancellationToken)
+    public async ValueTask ConnectAsync(ClientOpts connectOpts, CancellationToken cancellationToken)
     {
-#pragma warning disable CA2016
-#pragma warning disable VSTHRD103
-        if (!_semLock.Wait(0))
-#pragma warning restore VSTHRD103
-#pragma warning restore CA2016
-        {
-            return ConnectStateMachineAsync(false, connectOpts, cancellationToken);
-        }
-
-        if (_flushTask is { IsCompletedSuccessfully: false })
-        {
-            return ConnectStateMachineAsync(true, connectOpts, cancellationToken);
-        }
-
+        Interlocked.Add(ref _counter.PendingMessages, 1);
+        await _semLock.WaitAsync(cancellationToken).ConfigureAwait(false);
         try
         {
             if (_disposed)
             {
                 throw new ObjectDisposedException(nameof(CommandWriter));
             }
 
-            var success = false;
-            try
-            {
-                _protocolWriter.WriteConnect(connectOpts);
-                success = true;
-            }
-            finally
+            PipeWriter bw;
+            lock (_lock)
             {
-                EnqueueCommand(success);
+                if (_pipeWriter == null)
+                    ThrowOnDisconnected();
+                bw = _pipeWriter!;
             }
+
+            _protocolWriter.WriteConnect(bw, connectOpts!);
+            await bw.FlushAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
_pipeWriter is replaced on reconnect in `Reset()`. I wanted to avoid any possible stale reads. afaik in this case semaphore or channel 'async locks' would provide any memory consistency guarantees.